[css-grid] Consider sum of non-orthogonal nested subgrids' margin,

sammygill opened the pull request:
#### 94edbaabbf368aa6ffa9a0d21601c668a0ba095a
<pre>
[css-grid] Consider sum of non-orthogonal nested subgrids&apos; margin,
border,  and padding for first baseline alignment in the column axis.
<a href="https://bugs.webkit.org/show_bug.cgi?id=262605">https://bugs.webkit.org/show_bug.cgi?id=262605</a>
rdar://problem/116443747

Reviewed by Matt Woodrow.

To compute the extra layer of margin for subridded items in this
scenario, we will simply walk up the subgrid ancestors until we reach
the outermost parent grid. For each subgrid, we will take its mbp and
add it to a cumulative offset value. By the time we reach the outermost
parent grid we should have a sum of all of the nested subgrid margin,
border, and padding values.

If at any point we run into an orthogonal subgrid we will return 0 to
fallback to previous behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-008-expected.txt:
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::GridBaselineAlignment::ascentForChild const):

Canonical link: <a href="https://commits.webkit.org/268928@main">https://commits.webkit.org/268928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/335303fa0c8680c81730c3db7292755cfc1e5cdc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20944 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19510 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21182 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24588 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21525 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20781 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20927 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18188 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23689 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18088 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19002 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25287 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19174 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23201 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19759 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16789 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19011 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5050 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19584 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow reviewed (approved):
Looks good!

I assume `ascentForChild` (descendant, really) is only called for the 'root' grid in that axis?

Doesn't seem like there's any way to verify that we're not walking past the ancestor that called us, which would return an invalid result. I guess we definitely shouldn't be using that results for anything either though.

sammygill committed:
[css-grid] Consider sum of non-orthogonal nested subgrids' margin,
border,  and padding for first baseline alignment in the column axis.
https://bugs.webkit.org/show_bug.cgi?id=262605
rdar://problem/116443747

Reviewed by Matt Woodrow.

To compute the extra layer of margin for subridded items in this
scenario, we will simply walk up the subgrid ancestors until we reach
the outermost parent grid. For each subgrid, we will take its mbp and
add it to a cumulative offset value. By the time we reach the outermost
parent grid we should have a sum of all of the nested subgrid margin,
border, and padding values.

If at any point we run into an orthogonal subgrid we will return 0 to
fallback to previous behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-008-expected.txt:
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::GridBaselineAlignment::ascentForChild const):

Canonical link: https://commits.webkit.org/268928@main
